Source/Applications/SystemCenter/Controllers/SystemCenter/SystemCenterController.cs

```diff
@@ -744,90 +747,166 @@ public IHttpActionResult Post([FromUri] string extension, [FromUri] string meter
+                {
+                    // Finding Phases that do not yet exist and adding them to the database
+                    Func<ParsedChannel, openXDA.Model.Phase, bool> phaseEquality =
+                        (ParsedChannel channel, openXDA.Model.Phase phase) => phase.Name == channel.Phase;
+                    Func<ParsedChannel, openXDA.Model.Phase> createPhase = 
+                        (ParsedChannel channel) =>
+                            new openXDA.Model.Phase()
+                            {
+                                Description = channel.Phase,
+                                Name = channel.Phase
+                            };
+                    CheckAndAddRecords(connection, channels, phaseEquality, createPhase);
+                    
+                    // Same as above but characteristic
+                    Func<ParsedChannel, openXDA.Model.MeasurementCharacteristic, bool> charEquality = 
+                        (ParsedChannel channel, openXDA.Model.MeasurementCharacteristic characteristic) => characteristic.Name == channel.MeasurementCharacteristic;
+                    Func<ParsedChannel, openXDA.Model.MeasurementCharacteristic> createChar =
+                        (ParsedChannel channel) =>
+                            new openXDA.Model.MeasurementCharacteristic()
+                            {
+                                Description = channel.MeasurementCharacteristic,
+                                Name = channel.MeasurementCharacteristic,
+                                Display = false
+                            };
+                    CheckAndAddRecords(connection, channels, charEquality, createChar);
+
+                    // Same as above but characteristic
+                    Func<ParsedChannel, openXDA.Model.MeasurementType, bool> typeEquality = 
+                        (ParsedChannel channel, openXDA.Model.MeasurementType type) => type.Name == channel.MeasurementType;
+                    Func<ParsedChannel, openXDA.Model.MeasurementType> createType =
+                        (ParsedChannel channel) =>
+                            new openXDA.Model.MeasurementType()
+                            {
+                                Description = channel.MeasurementType,
+                                Name = channel.MeasurementType
+                            };
+                    CheckAndAddRecords(connection, channels, typeEquality, createType);
+                }
+
             return Ok(channels);
         }
+
+        private void CheckAndAddRecords<T>(AdoDataConnection connection, IEnumerable<ParsedChannel> channels, Func<ParsedChannel, T, bool> equalityFunction, Func<ParsedChannel, T> newRecordFunction) where T : class, new ()
+        {
+            TableOperations<T> recordTable = new TableOperations<T>(connection);
+            List<T> allRecords = recordTable.QueryRecordsWhere("1=1").ToList();
```

Comment:
This gives us datarows rather than objects of our given type. We could convert these to the object but thenthe code is more complicated than this one liner.

Reply:
It should be calling this overload, which returns `IEnumerable<T>`.
https://github.com/GridProtectionAlliance/gsf/blob/ed39aeaba8e99b026314cad27ce22e51fcf5f8d5/Source/Libraries/GSF.Core/Data/Model/TableOperations.cs#L686